Add selection modes - "Select All" and "Select Low Priority"

dragunoff opened the pull request:
This is a proposed implementation of the modifiers discussed in #14694. It adds two modes that alter the priority calculation. The modes are activated by holding down modifiers keys when selecting by mouse dragging.

~~Two new fields are added to the `Selectable` trait:~~
~~- `IncludeInSelectAllMode = true`~~
~~- `IncludeInSelectLowPriorityMode = false`~~

One new field is added to `Selectable`:
- `PriorityModifiers` with valid values `None` (default), `Ctrl` and `Alt`

Using the new field I have impleted two selection modes as YAML templates. This is how they work:
- **Flat Mode:** holding down <kbd>Ctrl</kbd> selects all actors that inherit `^FlatSelectionMode`. Buildings are not included but would still be selected if there are no higher priority actors. 
_Example use case_: Retreating all of your units (combat and non-combat).
_Example use case_: Moving together a mixed group of units (combat and non-combat).
- **Low Priority Mode:** holding down <kbd>Alt</kbd> selects all actors that inherit `^LowPrioritySelectionMode`. This is added to all non-combat units.
_Example use case_: Retreating non-combat units (harvesters or engies for example).
_Example use case_: Selecting low priority units among blobs (engies, harvesters, MCV).

![изображение](https://user-images.githubusercontent.com/1355810/57387167-7f997880-71be-11e9-93df-c9262cd12fc9.png)

Remarks:
- I was thinking about adding a third modifier - <kbd>Ctrl</kbd> + <kbd>Alt</kbd> but decided that this is good enough for now
- This can be used by modders to set up their own selection modes ("Aircraft Mode" for example)
- It doesn't work with hotkey selection (hotkeys with modifiers are considered a different hotkey).
- Possible hotkey implementation will conflict with the default bindings for bookmarks (<kbd>Ctrl</kbd>+<kbd>q</kbd> and <kbd>Alt</kbd>+<kbd>q</kbd>).
- The selection modes are not exposed in any way in the UI. This should be addressed in another PR.

abcdefg30 commented on mods/ts/rules/defaults.yaml:
```diff
@@ -823,7 +824,6 @@
 	Targetable@AIRBORNE:
 		TargetTypes: Air
 		RequiresCondition: airborne
-	Selectable:
```
Is this correct? Travis shows some errors that could be related.

dragunoff replied:
No, that was not intentional. I did not notice it in the diff. Fixed.

matjaeck commented:
I think it's a good approach.

obrakmann commented on OpenRA.Game/SelectableExts.cs:
```diff
@@ -28,7 +28,18 @@ public static int SelectionPriority(this ActorInfo a)
+			if (selectAllMode && a.Info.TraitInfo<SelectableInfo>().IncludeInSelectAllMode)
+				basePriority = 10;
+
+			if (selectLowPriorityMode && !a.Info.TraitInfo<SelectableInfo>().IncludeInSelectLowPriorityMode)
```
at this point it probably makes sense to store `SelectableInfo` in a variable

dragunoff replied:
Thanks for the suggestion. Implemented.

dragunoff commented:
Rebased and updated.

dragunoff commented:
Would this need an upgrade rule for 3rd party mods and maps?

pchote commented on OpenRA.Game/Traits/Selectable.cs:
```diff
@@ -11,11 +11,17 @@
 namespace OpenRA.Traits
 {
-	[Desc("This actor is selectable. Defines bounds of selectable area, selection class and selection priority.")]
+	[Desc("This actor is selectable. Defines bounds of selectable area, selection class and selection priority. Defines wether actor is included in selection modes.")]
```
whether.

dragunoff replied:
Fixed.

pchote commented:
> Would this need an upgrade rule for 3rd party mods and maps?

This isn't changing any default behaviour, right? If so, then no.

dragunoff commented:
> 
> 
> > Would this need an upgrade rule for 3rd party mods and maps?
> 
> This isn't changing any default behaviour, right? If so, then no.

It's not changing the default behavior. It's just that the new feature won't be very usable without the YAML setup.

pchote commented on OpenRA.Game/Traits/Selectable.cs:
```diff
@@ -11,11 +11,17 @@
+		public readonly bool IncludeInSelectAllMode = true;
+
+		[Desc("Include actor in low priority selection mode.")]
+		public readonly bool IncludeInSelectLowPriorityMode = false;
```
Perhaps a personal style nit, but wouldn't it be better to expose these as flags?

```csharp
[Flags]
public enum SelectionPriorityModifiers
{
	All = 1,
	LowPriority = 2
}

[Desc("Allow selection priority to be modified using a hotkey.",
	"Valid values are All (raise priority when Ctrl pressed) and",
	"LowPriority (lower priority when Alt pressed).")]
public readonly SelectionPriorityModifiers PriorityModifiers = SelectionPriorityModifiers.All;
```

dragunoff replied:
I've changed the implementation to use flags.

pchote reviewed (commented):
Behaves well in game and code looks mostly fine. Just a couple of requests:

dragunoff commented on OpenRA.Game/SelectableExts.cs:
```diff
@@ -28,7 +28,18 @@ public static int SelectionPriority(this ActorInfo a)
 		public static int SelectionPriority(this Actor a)
 		{
-			var basePriority = a.Info.TraitInfo<SelectableInfo>().Priority;
+			var modifiers = Game.GetModifierKeys();
```
Implemented.

pchote commented on OpenRA.Game/SelectableExts.cs:
```diff
@@ -45,37 +47,56 @@ public static int SelectionPriority(this Actor a)
 		}
 
-		public static Actor WithHighestSelectionPriority(this IEnumerable<ActorBoundsPair> actors, int2 selectionPixel)
+		static int BaseSelectionPirority(SelectableInfo info, Modifiers modifiers)
```
Typo: `Pirority`

pchote commented on OpenRA.Game/SelectableExts.cs:
```diff
@@ -45,37 +47,56 @@ public static int SelectionPriority(this Actor a)
+			var lowPriorityMode = modifiers.HasModifier(Modifiers.Alt);
+			var priority = info.Priority;
+
+			if (flatPriorityMode) {
```
Style nit: Opening braces go on their own line.

reaperrr commented on OpenRA.Game/Traits/Selectable.cs:
```diff
@@ -9,13 +9,29 @@
+			"Valid values are None (priority is not affected by modifiers)",
+			"FlatPriority (priority is set to 10 when Ctrl pressed) and",
+			"LowPriority (priority of actors without this value is set to 0 when Alt pressed).")]
+		public readonly SelectionPriorityModes SelectionPriorityModes = SelectionPriorityModes.FlatPriority;
```
I might be missing something here, but doesn't FlatPriority change the default behavior compared to bleed? Shouldn't this be None instead?

dragunoff replied:
Either of the values here changes the default behavior when a modifier key (<kbd>Ctrl</kbd> or <kbd>Alt</kbd>) is pressed. Even if it's `None` the priority would be set to `0` when <kbd>Ctrl</kbd> is pressed during a selection. 

Perhaps the description here is not accurate. Actually I've been struggling to come up with good names and description for this feature. Because even though the modes are called "flat" and "low priority" they are not tied to the priority of an actor. They may as well be called "select ctrl" and "select alt"...

dragunoff replied:
A value of `None` probably makes more sense as a default. I didn't do it like that because most of the actors are included in the "flat" mode and I wanted to spare some YAML lines...

pchote replied:
IMO the difficulty in describing this probably reflects that this isn't exposed to the yaml in the best way.

What if we changed the behaviour such that `None`, the default, keeps things as they are on bleed. Then expose usefully named templates that can be inherited, e.g. `Inherits@SELECTIONPRIORITY: ^SupportActorSelectionPriority`?

dragunoff replied:
I agree that it's probably not the best way to expose it in YAML. I am open to suggestions. However I don't understand the template and inheritance idea -- could you provide an example or point me to a similar implementation of another feature?

pchote replied:
The `^AutoTarget*` templates that most actors implement.

pchote commented:
What's the status here @dragunoff?

dragunoff commented:
@pchote I definitely want to finish this. But it turned out to be more complex than I thought. I think this warrants some more discussion.

The problem for me is that the modes are not tied to the selection priority, so calling them "flat" and "low priority" doesn't make sense... I thought about just calling them "default", "mode 1", "mode 2", "mode 3" and then let actors have different priority for each mode (falling back to the default priority). We can even let mods define the names for these modes and if the want to even use them. That would make the feature very flexible. 
```
# A cobmat unit, priority lowered in MODE2 and MODE3
Priority: 10
Priority@MODE1: 10
Priority@MODE2: 0

# A support unit, priority raised in MODE1
Priority: 6
Priority@MODE1: 10

# A high priority combat unit, priority raised in MODE3
Priority: 10
Priority@MODE3: 20
```

I'm not sure about the YAML notation, just trying to explain the idea.

pchote commented:
How important is the "low priority" modifier? Can we get most of the gameplay benefits without the complexity by ignoring all the per-actor definitions and implementing this as a global "ignore priority" modifier during selection?

dragunoff commented:
Yes, ignoring priority is fine - that way we can keep the normal priority based selection and introduce modified selections that force raise/lower the priority. The naming problem remains though.

pchote commented:
What is there to name at this point? We wouldn't need to expose anything to the mod rules.

dragunoff commented:
> What is there to name at this point? We wouldn't need to expose anything to the mod rules.

How are we going to then define which actors should be included in which selection? 🤔

I may be misunderstanding your idea...

pchote commented:
My suggestion was to have it include all actors that are controllable by the player.

If we really do need a distinction for buildings we could call have something like "ValidForFlattenedPriority: (true|false)".

dragunoff commented:
Ah, I see - you suggest dropping the "low priority" mode entirely. I'd really like to make it work though ☺️ And the naming problem won't go away - what would stop modders from repurposing the "flat" mode into something else? That may not be a huge problem but it bugs me...

pchote commented:
My motivation with exploring these other options is so that we can find a way to get this merged in time for the playtest branching.

dragunoff committed:
Allow selection priority to be modified using a hotkey

dragunoff committed:
Add "Flat" and "Low Priority" selection modes to default mods

dragunoff commented:
Okay, so I have simplified this a little bit. And I'm finally (mostly) satisfied with the result. 

I have dropped the selection modes names from the C# code and just named the flags after the actual modifier keys (`Ctrl` and `Alt`). This solves the naming problem and gives modders the freedom to use the feature however they want. For the default mods I've created two selection modes that are implemented as YAML templates which actors inherit. 

I've also dropped my original idea for the "Low Priority" mode (where selection priority was honored) - now modifiers simply raise the selection priority to `int.MaxValue` thus always flattening the priority for affected actors.

dragunoff commented:
Rebased on latest bleed.

abcdefg30 reviewed (commented):
Seems like you missed engineers (and thumpers?) in D2k. Looks good to me otherwise.

dragunoff commented:
> 
> 
> Seems like you missed engineers (and thumpers?) in D2k. Looks good to me otherwise.

It is intentional as they don't have a lower selection priority in D2k. This should be tackled as part of #16662

teinarss reviewed (approved):
Tested in all the mods and the code looks fine

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f3e83db7bd88d08dde9ea0a5557e3b3379a25611...d2b5c48c2528d629acc8ef41fffe229ac9c63613
